### ISSUE TITLE
ACCESS-CM3 Builder

### ISSUE THREAD

**Issue by dougiesquire** [ACCESS-NRI/access-nri-intake-catalog#168]
ACCESS-CM3 is now at the point of producing output. It could be useful to have an ACCESS-CM3 Builder to allow us to build datastores for ACCESS-CM3 data.

ping @MartinDix, @kieranricardo

**Comment by anton-seaice** [comment]
@sfiddes would also like a builder for UM regional nesting suite output - its not clear to be if this is the same or a different builder ? If someone is looking at this soon she can provide some example output :)

**Comment by marc-white** [comment]
I think I'm on the hook for new builders @anton-seaice , so if you can provide some demo data I can start looking into it.

**Comment by dougiesquire** [comment]
@paolap set up a builder (in a fork?) for Aus2200 output that could do what we need. Paola, are you planning on contributing that in to the main package?

**Comment by sfiddes** [comment]
I have put a couple of files here: /scratch/public/slf563/UM_RNS. It is fairly niche sorry, but if there is something that works for AUS2022 that would be a good start... Also, I've just used iris to convert from pp files, so the naming might also be a bit off. I am looking for a better way to do the data processing to use CF conventions etc.. So perhaps not the best dataset to start with - AUS2022 probably better, but with recognition that not everyone uses AUS2022 :)

**Comment by paolap** [comment]
> @paolap set up a builder (in a fork?) for Aus2200 output that could do what we need. Paola, are you planning on contributing that in to the main package?

Yes that was the idea, I was waiting to have time to review what I did and make it more flexible so that a user could control which extra fields would appear in the catalogue attributes.  I'm busy all this week and I would need to adapt it to whatever change Marc introduced,  I can have another look at it next week and than set a pull_request I did put  a link in another issue.

I also had a chat with Sonya previously about post-processing her data as Mopper has already the capacity of doing so, but currently starting from the raw model output rather than from the iris version. One issue with iris is that seems to make up standard_names, which is a potential source of information for mopper, so it would confuse the tool. Again I won't have anytime to look at this until next week.
Finally if there are examples of CM3 output I would love to get a sample so we can provide mapping for this version in our tool.

**Comment by marc-white** [comment]
@paolap how did you go with your new builder?

**Comment by paolap** [comment]
I didn't do anything more, maybe next week I'll have some time to review it and pull your changes. It was working for me for the example I wanted to run. The catalogue I produced is the one Navid used in his atmospheric cookbook example.

**Comment by paolap** [comment]
Finally, I'm going to have some time next week to look into this again, so far I just updated my fork and merge your updates from main. Let me know, if there's more from other branches that I should be aware of, i.e. more changes to the builders in particular.

**Comment by marc-white** [comment]
Hi all, is there some representative ACCESS-CM3 data I can use for this?

**Comment by paolap** [comment]
@marc-white finally I had sometime to look at this, way later than I would have liked to! I updated the MopperBuilder class to follow the changes you implemented including using the new class _AccessNCFileInfo

As mopper is a wrapper to CMOR that writes one variable per file it would be great if it was possible to avoid a multivariable file setup when it's not needed. Multivariable esm catalogues have some limitations, it's also potentially more confusing from a user point of view.

My changes are in my own fork in the aus2200 branch, the main branch should be up to date with the official repo:

https://github.com/paolap/access-nri-intake-catalog

**Comment by marc-white** [comment]
Thanks @paolap , I'm going to pull the branch into the main repo to take a look & continue work.

**Comment by marc-white** [comment]
@paolap what are you providing as the arguments `fpattern` and `toselect` to the `MopperBuilder` (as an example)?

**Comment by paolap** [comment]
This config file should show what they are for the dataset I sued as test:

https://github.com/paolap/access-nri-intake-catalog/blob/aus2200/config/access-mopper.yaml

Basically "fpattern" is set in the file used as configuration for mopper to build the directory structure of the post-processed output.
See:

"path_template" and "file_templat"e in this yaml file:
https://github.com/ACCESS-Community-Hub/ACCESS-MOPPeR/blob/main/ACDD_conf.yaml
~ line 38

This follow more strict rules for CMIP6 data as shown in the equivalent file if following CMIP6.
CMOR will use this template to create directories and build the filenames.

This is why it made sense for me to use the same setup for the builder. Then I added toselect to allow a user to chose what attributes they wanted in the catalogue.
NB that the fpattern won't be exactly the same, as for example the variable name is identified as variable_id in CMOR (can't be changed) but is identify as "variable" in your catalogue. But it's easy enough for a user to adapt the terms.

The date_range is determined by the frequency automatically by CMOR as fromdate-todate, hence it doesn't appear in the template.
For example for monthly data will be something like YYYYMM-YYYYMM
But for 3 hourly is YYYYMMDDhhmm-YYYYMMDDhhmm
Where YYYYMMDDhhmm will represent the actual timestamp of the first and last timestep.

**Comment by marc-white** [comment]
Hi @paolap , I think I'm making good progress on this. One question for you: I noticed you were trying to pull a `tracking_id` out of the files, and assign it to `file_id` within the catalog. What does this `tracking_id` do? Does it group files together to form a coherent experiment, or is it recording something else?

**Comment by marc-white** [comment]
@paolap I now have a `MopperBuilder` that I'd like to deploy in anger against a few full experiments. Would you be able to point me to some good data available on Gadi for that purpose?

**Comment by anton-seaice** [comment]
I am pretty sure Paola has finished up with Clex and probably won't be checking this anymore !

**Comment by marc-white** [comment]
Oh right.... damn. Any suggestions as to whom I should bother instead @anton-seaice ?

**Comment by anton-seaice** [comment]
Not really - @MartinDix or @kieranricardo or maybe Chermelle ?

**Comment by marc-white** [comment]
@MartinDix or @kieranricardo , would you be able to point me towards some full experiments that I can test this request against?

**Comment by marc-white** [comment]
@MartinDix and @kieranricardo , any thoughts on what data I could use to test this particular Builder?

**Comment by kieranricardo** [comment]
hi @marc-white, we have a 100+ yr run here: `/home/547/kr4383/cylc-run/cm3-run-29-01-2025-exp-runoff-smoothing-rmax-500-efold-1000`. 

The data locations are a little clunky at the moment, but you can find: 
- atmospheric data in `share/data/History_Data/netCDF`
- sea-ice data under `work/YYYYMM01/atmos/access-cm3-cice.h.YYYY-MM.nc`
- ocean data in `work/YYYYMM01/atmos/access-cm3-mom6.h.YYYY-MM.nc` (there's also` ...mom6.h.native_YYYY-MM.nc` and ` ...mom6.h.sfc_YYYY-MM.nc` files)

Let me know if there's anything I can help with!

**Comment by marc-white** [comment]
@kieranricardo unfortunately, I don't have access to that directory (it's your home). Can you either relax the permissions to allow group or 'other' read, or move the data to somewhere on `/g/data`?

**Comment by kieranricardo** [comment]
@marc-white woops meant to send the scratch location, you can find it here: `/scratch/tm70/kr4383/cylc-run/cm3-run-29-01-2025-exp-runoff-smoothing-rmax-500-efold-1000`

**Comment by marc-white** [comment]
@kieranricardo , awesome, I can see those data.

Re: the data locations being 'clunky' - is there a target/standard data location format that you'll be aiming for, or (as @paolap said above), it's likely to be more down to the individual preference of the person running the model?

**Comment by kieranricardo** [comment]
@marc-white yep we've got plans to have a more usable data format! Roughly, we're planning to:
- have the MOM and CICE data in the same format as OM3
- leave the UM data as is (so is in the same/similar format as ESM1.6)

And we'll be moving the data from run into the new format soon.

**Comment by kieranricardo** [comment]
@blimlim

**Comment by marc-white** [comment]
@kieranricardo I'm guessing the data to be ingested is that found in `share/data/History_Data/netCDF`?

**Comment by marc-white** [comment]
@blimlim awesome thanks!

**Comment by paolap** [comment]
@marc-white  just accessing this after a long break from work, apologies I lost access to my utas email a lot faster than expected.  If you want to test that the data I mentioned in bs94, any of these runs would do as they were produced with ACCESS-MOPPER, said so it miht not matter anymore as I think the nRI might not support the tool in the end, or at least that's what I heard. I'm really not working with ACCESS data anymore.

**Comment by marc-white** [comment]
Handing this one off to @charles-turner-1 (in the first instance).

I've updated the 186 branch to match `main`. I had to make some alterations to the way the Builders work to handle the odd file/directory structure of these data; however, this issue (and hence the need for those changes) will almost certainly go away once #432 is completed. Therefore, you'll probably just want to throw away this branch and start again, but at least all the test data & their expected outputs are kept here for now.

**Comment by chrisb13** [comment]
@charles-turner-1, I found the issue!

**Comment by charles-turner-1** [comment]
Closed by #498
